File: assignments/haskell/accumulate/accumulate_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import Accumulate (accumulate)
import System.Exit (ExitCode(..), exitWith)
import Data.Char (toUpper)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
       [ TestList accumulateTests ]

square :: Int -> Int
square x = x * x

accumulateTests :: [Test]
accumulateTests =
  [ testCase "empty accumulation" $
    [] @=? accumulate square []
  , testCase "accumulate squares" $
    [1, 4, 9] @=? accumulate square [1, 2, 3]
  , testCase "accumulate upcases" $
    ["HELLO", "WORLD"] @=? accumulate (map toUpper) ["hello", "world"]
  , testCase "accumulate reversed strings" $
    ["eht", "kciuq", "nworb", "xof", "cte"] @=?
    accumulate reverse ["the", "quick", "brown", "fox", "etc"]
  , testCase "accumulate recursively" $
    [["a1", "a2", "a3"], ["b1", "b2", "b3"], ["c1", "c2", "c3"]] @=?
    accumulate (\c -> accumulate ((c:) . show) ([1, 2, 3] :: [Int])) "abc"
  ]

File: assignments/clojure/accumulate/accumulate_test.clj
(ns accumulate.test (:require [clojure.test :refer :all]))
(load-file "accumulate.clj")

(defn- square [n] (* n n))
(defn- to-s [xs] (apply str xs))

(deftest empty-accumulation
  (is (= [] (accum/accumulate square []))))

(deftest accumulate-squares
  (is (= [1 4 9] (accum/accumulate square [1 2 3]))))
   
(deftest accumulate-upcases
  (is (= ["HELLO", "WORLD"] 
         (map to-s (accum/accumulate clojure.string/upper-case ["hello" "world"])))))

(deftest accumulate-reversed-strings
  (is (= ["eht" "kciuq" "nworb" "xof" "cte"]
         (map to-s (accum/accumulate reverse ["the" "quick" "brown" "fox" "etc"])))))

(deftest accumulate-recursively
  (is (= [["a1" "a2" "a3"] ["b1" "b2" "b3"] ["c1" "c2" "c3"]]
         (accum/accumulate #(accum/accumulate (fn [n] (str % n)) [1 2 3]) "abc"))))
 
(run-tests)

